### DISCUSSION TITLE
Decorator transforming a mandatory keyword arg into an optional keyword arg

### DISCUSSION THREAD

**Discussion by b4stien** [microsoft/pyright#7596] [Q&A]
I have many functions which share the same structure and base signature:

```python
async def get_stuff_from_external_api(*, client: Client | None, id: int) -> Resource:
    if not client:
        client = make_client()
    return Resource(await client.get(id))
```

To DRY my code I extracted this logic into a decorator, both the decorator and the functions now being:

```python
def external_api_decorator(func):
    @functools.wraps(func)
    async def wrapped_func(*args, **kwargs):
        client = kwargs.pop(client, None)
        if not client:
            client = make_client()
        return await func(*args, client=client, **kwargs)
    return wrapped_func

@external_api_decorator
async def get_stuff_from_external_api(*, client: Client, id: int) -> Resource:
    return Resource(await client.get(id))
```

My problem is now that I can't properly type `external_api_decorator` to keep both decorated functions initial parameters (`id` in my example) and returned value (`Resource` in my example).

I have tried many solutions (`Protocol`, `Generic`, `ParamSpec`, ...) but I think I'm stuck on `Concatenate`/`ParamSpec` not supporting keyword arguments. Am I missing something? It looks like a super classic use-case.

Here is [a pyright playground to reproduce](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=JYWwDg9gTgLgBAMwK4DsDGMIQDYGcCwAUEWtgIa65wBKAprhElGrQBQCUAXEXL3ACa0EcAPojgKYDDGtctbAgA0cKJzi4YULjz66wFAsUKkDcAMLZgtFDA7dCuigE90AoXADmtW3IXLg-GoSMOxwALQAfOqa9rq6UN5MKHAARAhYKUREgsIgZADWtCKkVjYc4VEWpTCxfAkwSeaW1rbsWYQ5cLQAHjC0UChk2CJkYMAigmjQZJhQrMjo2g58AAILGFh4AHQA7lCjuPOoaG3LvM6unXujYLT8IuusAFRkUB64yk9P%2BTuv70txPglFpwAC8cB%2Bf1wW0gYFYwJsygAchAULRToDeMBhCgIPAETUdJigc0bGC4HlCsVSa0iXF6o0yL8pIhjs8ocoCaCCZ9vr83rgMXVEgM4NcwLd7ut2isen0BkMRmMJrQpvtZkQLmg3MIvNINEgEAgHlAICARHL%2BoNhqNgM9OTS1FUWv5AnBgqFIjR6IxmLRarwGaK6AwmCxWEyyCyCVs9awAuwhbxNbgXNrOnkJOUvSi0QG4ABiczQBIYOBoCi0Olq0vSKDkyMsvUiA1Gk1mi29K2K23x-iggCMACYk4W4ABlAAWjGw-BUtAAbrQhqkQ76WJkzvOl4qYE5bvCS6q66O6UWAJLoI9lityZS4adIWcqKNyOBkOB725dKCmqB0iQa2PER63BRt4GbVtjQQU1zUtBUbTGPtQRSAAjV4UnYIA).

Does anyone know how I can type this? Am I stuck with "`Concatenate` and `ParamSpec` do not support keyword arguments"?

**Comment by Viicos** [comment]
> Am I stuck with "`Concatenate` and `ParamSpec` do not support keyword arguments"?

Unfortunately yes, see [this section](https://peps.python.org/pep-0612/#concatenating-keyword-parameters) of the PEP.

Maybe something like this, using a positional argument instead? (may need some tweaking to work with async):

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&locale=en-us&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAAqXkQDKCxAxjQMJgpvkzEU-Yl3IAbMeQBGYkXXAwwbMGICwAKA1tJAZx1ROYpIJgAuDVEtQE5PRo20oAXjoNmrNgAoARLW8BKDQAlZ0ISChAfIO8aZQA3SiRyFBgnAhAAV2JAzXVtW30AQVoASVoFJRVPcrBFZTEAbVoaIIBdf3N1KygAE2JgKAB9Qb4JYc8dYjFgWKMTUwM5lKgAHygAOR5iUM2UOQAqSjQdBdoAOiOdGn39gGsAd0vTs4fL-ygAWgA%2BKCCFs4B9lyfQGxAAHgIQEIxINyMhBn1lCB%2BLhPMAMrwFpxxJIZMQGtxeMIhAICUsYDRaK0WtSoAB6d7fKDFMoVepNGmdboAs5A-J6X7EHRgDIgNjELlWGx2XIAAXBkOhsPhiNwKJAGhBUDQxBggx0MAywGAg1AkEGCuIUPEyqQnm0xhSWPJ1xoSB6C1QMEZPyCQpFYolFisPKBIGIcWINvgrE8Or1BqNJrNEAtEKtSrhSDOw1GMMG-iAA)

```python
from typing import TypeVar, ParamSpec, Concatenate, Callable, Protocol

class Client:
    pass

P = ParamSpec("P")
R = TypeVar("R", covariant=True)

class APIProtocol(Protocol[P, R]):
    def __call__(self, client: Client | None = None, *args: P.args, **kwargs: P.kwargs) -> R: ...


def external_api_decorator(func: Callable[Concatenate[Client, P], R], /) -> APIProtocol[P, R]:
    ...

class Resource:
    pass

@external_api_decorator
def get_stuff_from_external_api(client: Client, *, id: int) -> Resource:
    ...

reveal_type(get_stuff_from_external_api.__call__)
```
